### ISSUE TITLE
Approximate faces in surface coordinates

### ISSUE THREAD

**Issue by hannobraun** [hannobraun/Fornjot#568]
When approximating a face, the points that make up the approximation are generated as 3D points. This is a bit nonsensical, as approximation could -theoretically- just as well be done in 2D surface coordinates, and those 3D points are immediately converted to 2D coordinates anyway, for the triangulation.

This has been fine for a while, but now I discovered a bug in the 3D to 2D conversion logic: https://github.com/hannobraun/Fornjot/issues/567

And in fact, converting 3D approximations into 2D seems to be the only use case for that broken code.

As best I can tell, #567 would be non-trivial to fix and might cause a loss in generality (`SweptCurve` would likely need to be replaced with separate structs for planes and cylinders, which would negatively affect our ability to support more curve types in the future). Therefore, a different solution could be to just remove the need for that code existing in the first place, by creating approximations in 2D, avoiding the detour through 3D.

This would also have positive effects on #250, which could become much easier to implement. On the negative side, this would add some complexity to the representations of edges and faces, which would have to keep track of their bounding vertices/edges in their local coordinate system.

I'll try to go ahead and implement this anyway, as I believe it would result in cleaner code overall, in addition to solving the immediate problem (#567). There might be some problems that I'm not seeing right now though. We'll see.

**Comment by hannobraun** [comment]
I've made some good progress towards achieving this, but getting this over the finish line will take some more effort. The problem is that the next step is going to require careful modification of quite a lot of code, which is not only tedious, but error-prone. And most of that code shouldn't exist in the first place. Its existence reflects a lack of suitable methods of the `Shape` API.

#399 is probably a partial solution to this problem, but there are other new APIs that could help here (I've left some thoughts in the commits in #544).

I will look into this now, as solving this issue (and replacing all that problematic code) should make further progress here much easier.

**Comment by hannobraun** [comment]
I've removed all the blockers, and am back to working on this issue directly.

**Comment by hannobraun** [comment]
I have more progress towards this issue in local branches, but there's a bug in there that I haven't tracked down yet. I'm now working on #613, which should make the source of that bug quite obvious.

**Comment by hannobraun** [comment]
#613 is taken care of (for now), and I'm back to working on this issue directly.

**Comment by J-F-Liu** [comment]
If the purpose of 3D to 2D conversion is to do triangulation, is it possible and better to do triangulation on approximated 3D points without converting back to 2D surface coordinates?

**Comment by hannobraun** [comment]
Thank you for the suggestion, @J-F-Liu! I had considered this previously, and I see two problems; one immediate, one long-term.

The immediate problem is, that the library we're using for triangulation ([spade](https://crates.io/crates/spade)) is working with 2D coordinates. This means, we'd either have to replace the triangulation code with something else, or use some hack to make it work. Either way, it's a problem.

Long-term, I think trying to triangulate 3D coordinates would be a dead-end. Imagine how triangulation of a curved surface would work. Using 2D surface coordinates, the surface would be responsible to figure out how many points are required for a sufficient approximation, and where those points should be located. The triangulation algorithm would receive those 2D points, create the triangles, and convert those 2D triangles back into a 3D mesh.

Now imagine what would happen, if the triangulation algorithm would receive 3D points to approximate the surface. Consider a surface that looks something like this (sectional side view shown):
![surface](https://user-images.githubusercontent.com/85732/173036317-383b6a2c-e823-4537-b434-3806366d6b4a.png)


The crosses along the line are the approximated points that the surface delivers.

Consider how the approximation algorithm might decide to connect those points, based on proximity. I've drawn some connections here (in purple) that would be plain wrong, as they would produce triangles inside of the shape, not along the surface:
![surface-with-wrong-lines](https://user-images.githubusercontent.com/85732/173036498-a821f077-c33f-410b-b92f-7b84241617b3.png)

I guess there could be some heuristic to determine which lines are valid selections and which aren't, based on the maximum distance of the line to the surface or something like that, but that sounds like a hacky way to do it. I wouldn't be surprised if that would lead to very interesting bugs.

If the triangulation receives points in surface coordinates, on the other hand, this won't be a problem. The Delaunay triangulation will just do the right thing naturally, without knowing or caring about what the surface looks like.

**Comment by hannobraun** [comment]
After not making a lot of progress on this for the last 2 weeks or so, I've recently made a breakthrough with #680. I started building upon this breakthrough today, trying to start wrapping up this issue. But while doing so, I've had another breakthrough of sorts: I've been thinking on how to simplify aspects of the kernel recently, and some of those thoughts have today solidified into a concrete plan. I've written that plan down in #691.

I've decided to work on #691 immediately, and put this issue aside while I'm doing that. If I can actually realize those simplifications, that will make everything else (including this issue) easier, so I think it's the right decision.
